### ISSUE TITLE
Truthiness of None Does not Narrow Optional Types

### ISSUE THREAD

**Issue by WillAyd** [python/mypy#7249]
Note: if you are reporting a wrong signature of a function or a class in
the standard library, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues

Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request?
   I believe this is a bug

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```python
from typing import Optional

def the_method(maybe_int: Optional[int]):
    can_divide = bool(maybe_int)
    if can_divide:
        100 / maybe_int
```

* What is the actual behavior/output?
```sh
error: Unsupported operand types for / ("int" and "None")
```

* What is the behavior/output you expect?
   I would think that the conditional branch would recognize that only `int` values would ever make it there

* What are the versions of mypy and Python you are using?
mypy 0.720 with Python 3.7.3

* Do you see the same issue after installing mypy from Git master?
Yes still exists on master

**Comment by refi64** [comment]
If you change it to just `if maybe_int:` then it type-checks successfully.

**Comment by WillAyd** [comment]
That's correct - this is certainly a toy example. However this came from an actual application where the condition was more like `can_divide = maybe_int and <something_else>`. 

Regardless of what the remaining conditions are it seems possible to statically infer that `None` values would not make their way into the conditional branch

**Comment by JukkaL** [comment]
This is a known limitation. We might want to make mypy understand code like this better at some point, if we can come up with a simple enough implementation. It doesn't seem very high priority since you can work around these issues with a cast or an assert, and we don't know how frequently this happens in typical Python code.

Here's a sketch how this might be implemented (I haven't thought this through carefully):

1. Recognize certain certain assignments such as `x = bool(y)` and `x = y is not None` and maybe `x = isinstance(y, str)` as "type check aliases". (Also things like `x = bool(y) and <whatever>`.)
2. If the target of such alias is used as a condition, treat it as equivalent to the original expression in terms of type inference.
3. If `x` or `y` gets reassigned, invalidate the alias in that scope.
